hera_cal/io.py

```diff
@@ -2106,21 +2150,38 @@ def write_vis(fname, data, lst_array, freq_array, antpos, time_array=None, flags
         integration_time = np.ones_like(time_array, dtype=np.float64) * np.median(np.diff(np.unique(time_array))) * 24 * 3600.
 
     # get data array
-    data_array = np.moveaxis(list(map(lambda p: list(map(lambda ap: data[str(p)][ap], antpairs)), pols)), 0, -1)
+    if data_array is None:
+        data_array = np.zeros((Nbls, Ntimes, Nfreqs, Npols), dtype=complex)
+        for i, antpair in enumerate(antpairs):
+            for j, pol in enumerate(pols):
+                data_array[i, :, :, j] = data[antpair + (str(pol),)]
 
     # resort time and baseline axes
     data_array = data_array.reshape(Nblts, 1, Nfreqs, Npols)
+    
     if nsamples is None:
         nsample_array = np.ones_like(data_array, float)
     else:
-        nsample_array = np.moveaxis(list(map(lambda p: list(map(lambda ap: nsamples[str(p)][ap], antpairs)), pols)), 0, -1)
+        if isinstance(nsamples, DataContainer):
+            nsample_array = np.zeros((Nbls, Ntimes, Nfreqs, Npols), dtype=float)
+            for i, antpair in enumerate(antpairs):
+                for j, pol in enumerate(pols):
+                    nsample_array[i, :, :, j] = nsamples[antpair + (str(pol),)]
+        else:
+            nsample_array = nsamples
```

Comment:
Why is this being handled differently? And why aren't we checking if it has the right shape (if it's not None)?

Reply:
Removed.